frontends/p4/defaultArguments.cpp

```diff
@@ -27,14 +27,18 @@ class TypeNameSubstitutionVisitor : public TypeVariableSubstitutionVisitor {
     explicit TypeNameSubstitutionVisitor(
         const TypeVariableSubstitution *bindings, const TypeMap* typeMap)
             : TypeVariableSubstitutionVisitor(bindings, true), typeMap(typeMap)
-    { CHECK_NULL(typeMap); setName("TypeNameSubstitution"); }
+    { CHECK_NULL(typeMap); setName("TypeNameSubstitution"); visitDagOnce = false; }
 
     const IR::Node* preorder(IR::Type_Name* tn) override {
         auto t = typeMap->getTypeType(getOriginal<IR::Type>(), true);
         if (auto tv = t->to<IR::ITypeVar>())
             return replacement(tv, tn)->to<IR::Type>()->getP4Type();
         return tn;
     }
+    // When cloning the value of an argument we want to make a fresh
+    // copy for each new InfInt type, and not share the same one.
+    const IR::Node* postorder(IR::Type_InfInt*) override
+    { return new IR::Type_InfInt(); }
```

Comment:
Curious: Why do we visit the default arg in the second statement as `Type_InfInt` but in the first statement as `Type_Bits`? This is just what I saw from IR dump

Reply:
In general the IR representation is a DAG, and not a graph; some nodes can share children. The node ID is unique, so in the IR dump look for nodes that have the same ID. In type inference InfInt is treated like a type variable, and unified with some other type to infer a width. That's why you want to create different type variables for each instantiation of the default value: so each can be unified independently. 

The DAG representation is subtle and has lead to several bugs so far. It is also related to the visitDagOnce flag of the visitor, which indicates whether shared nodes are visited only once or once for each parent. One has to keep in mind that the IR is a DAG, and whether shared nodes represent the "same" value really depends on the transformations that you are applying.